### ISSUE TITLE
Always unconditionally remove trailing white spaces

### ISSUE THREAD

**Issue by maximbaz** [mawww/kakoune#2175]
I noticed that every language in `rc` has to deal with trailing white spaces on its own, and `ag '(remove|clean).*trailing.*space' --stats` shows that there are currently 38 files that clean white spaces. This is a lot of duplicated logic! 

With the same motivation "kak is a code editor, and not arbitrary data editor" as mentioned in #2147, I propose to consider making kak always remove trailing white spaces. A simple snippet below should simplify those 38 files and also remove the handling of trailing white spaces from editorconfig parsing:

```
define-command trim-whitespaces -docstring 'trim trailing whitespaces' %{
    try %{ execute-keys -draft \%s\h+$<ret>d }
}

hook global BufWritePre .* %{ trim-whitespaces }
```

**Comment by danr** [comment]
I think this is blocked because there is no reliable way to track dependencies between scripts, see mawww's comment here: https://github.com/mawww/kakoune/pull/1811#issuecomment-360634833

**Comment by lenormf** [comment]
Actually the language support scripts I've written do not handle trailing whitespace, as I think it's not the indentation hooks' purpose to handle that.

Secondly, removing all trailing whitespace by default without the user's consent is also bad, the editor shouldn't assume how the user wants a buffer's contents to be formatted.

Example: I regularly have to fix code that wasn't written by me, that has trailing whitespace everywhere (and even LF and CRLF line endings mixed together), if the editor were to arbitrarily fix those issues, I would end up with patches that are 99% cosmetic changes. If the user wants proper formatting, they should use `:format` or declare hooks in their configuration.

Finally it's not an issue that some code is duplicated across a few files, as the user is free to load whatever script they want, and factorizing redundant code would make things harder to cherry-pick scripts in `autoload` (c.f. @danr's answer regarding dependencies).

**Comment by maximbaz** [comment]
There is probably no source code that requires trailing whitespaces to be preserved, `git diff -w` and modern code review tools are able to hide whitespace differences, so I don't think it's too bad if kakoune would slightly clean code files that you touch. It's a good practice to not have trailing whitespaces in the code, and cleaning them up is not the same as reformatting the file.

Having said that, I'm all for consistency. If kakoune unconditionally adds a missing terminating newline to the file and cleans up trailing whitespaces _sometimes_, it can just as well cleanup the trailing whitespaces all the time — less people in the world will commit code with such kind of issues (yay). Alternatively don't mess with trailing whitespaces in _all_ languages and mention on the wiki how to add the hook.

This is just a proposal, and thank you for commenting and providing great arguments. Good point about mixed LF and CRLF line endings, this is another case to consider. In the meantime, I want to investigate now how to write a hook to fix the mixed line endings 😉

**Comment by occivink** [comment]
I agree with @lenormf  that these hooks don't really belong with the language support scripts, and I'd be in favor of removing them. 
Right now doing it unconditionally when leaving insert mode is just too greedy, doing something as simple as pasting an identifier after inserting a newline is frustrating, and doesn't have an obvious workaround. 

I also think they're sufficiently trivial to let the users set them up as they want (on a binding, before saving or even automatically when leaving insert mode).

**Comment by mawww** [comment]
To give a bit of context, I believe the initial language support scripts I wrote do that for a simple reason: They need to cleanup the auto-inserted whitespaces.

When you open a newline, the previous line indent is copied, or maybe the indent gets increased due to something on the previous line. Those inserted whitespaces needs to be removed if we hit `<ret>` again, or if we quit insert mode with `<esc>`. The fact that they cleanup the existing text is a side effect, they are mostly there to make the auto-indent behaviour acceptable.

**Comment by maximbaz** [comment]
This is interesting, question is: do we still want this behavior? For example in python this doesn't happen (whitespaces don't get magically removed when I press `<ret>` again), and nobody seems to be bothered because of this. After all, `:format` or people's own hooks to cleanup trailing whitespaces (like I showed in the first message) will take care of them.

So far I have an impression that the consensus is going towards removing any post-formatting logic from syntax files. In my mind we should only keep the hooks that help you _type_, things like increase indent after typing `{` and decrease indent after typing `}`. But there should be no hooks that post factum reformat my code — if I added whitespaces and pressed `<ret>`, they should remain.

**Comment by mawww** [comment]
Not sure what you meant, if I open a new python file, type `def test():<ret><ret>foo` the first line of the function is empty, it does not contain trailing whitspaces. The hooks (IMHO correctly) cleaned up the indent it inserted.

**Comment by maximbaz** [comment]
I updated my message too late, sorry for confusion, yes it happens in python file as well. So I read this as you want to keep this behavior?

**Comment by vbauerster** [comment]
> There is probably no source code that requires trailing whitespaces to be preserved, `git diff -w` and modern code review tools are able to hide whitespace differences, so I don't think it's too bad if kakoune would slightly clean code files that you touch. It's a good practice to not have trailing whitespaces in the code, and cleaning them up is not the same as reformatting the file.
> 
> Having said that, I'm all for consistency. If kakoune unconditionally adds a missing terminating newline to the file and cleans up trailing whitespaces _sometimes_, it can just as well cleanup the trailing whitespaces all the time — less people in the world will commit code with such kind of issues (yay). Alternatively don't mess with trailing whitespaces in _all_ languages and mention on the wiki how to add the hook.
> 
> This is just a proposal, and thank you for commenting and providing great arguments. Good point about mixed LF and CRLF line endings, this is another case to consider. In the meantime, I want to investigate now how to write a hook to fix the mixed line endings 😉

It is valid case to have trailing whitespaces in `md` (markdown) files, which means explicit line break in resulting output. I had bad experience when was making a [PR](https://github.com/bahlo/go-styleguide/pull/11#discussion-diff-185476983L105) to a `md` file with "auto cleanup trailing whitespace" option on in my editor.

**Comment by eraserhd** [comment]
Trailing whitespace in multi-line strings could be intentional, also.  (And
multi-line strings are definitely language specific.)

**Comment by maximbaz** [comment]
I think the consensus now is to **not** touch whitespaces, also PR #2273 will fix this issue I think.

**Comment by occivink** [comment]
Just fyi if you don't like the automatic removal of whitespaces when exiting insert mode you can now use
`set global disabled_hooks .*-trim-indent`

**Comment by wcasanova** [comment]
is there any way to be able to see the spaces in this way?
![imagen](https://user-images.githubusercontent.com/12401985/95578912-19480380-0a0b-11eb-9b94-00fbbf0f3392.png)

**Comment by alexherbo2** [comment]
``` kak
add-highlighter global/show-trailing-whitespaces regex '\h+$' 0:Error
```

**Comment by AndrewVos** [comment]
@alexherbo2 I've tried that code snippet but it doesn't seem to do anything. Do I need to enable it somehow?

**Comment by clarfonthey** [comment]
Perhaps adding a common whitespace hook to #4266 might help with this. That way, filetypes could have default trimming behaviour but it could be overridden by editorconfig, or manually.

**Comment by ismay** [comment]
It took me a long time to figure out why kakoune was trimming whitespace in my markdown files, when I have an override for markdown in my `~/.editorconfig`:

```
# https://editorconfig.org

root = true

[*]
charset = utf-8
end_of_line = lf
indent_size = 2
indent_style = space
insert_final_newline = true
trim_trailing_whitespace = true

[*.md]
trim_trailing_whitespace = false

[*.fish]
indent_size = 4
indent_style = space

[**/i3blocks/config]
trim_trailing_whitespace = false

[COMMIT_EDITMSG]
max_line_length = 100
```

Only _after_ adding this line to my `.kakrc`:

```
set global disabled_hooks .*-trim-indent
```

does kakoune respect my .editorconfig  `trim_trailing_whitespace` settings for markdown files. The confusing thing is, kakoune already respected my .editorconfig disabling of trailing whitespace trimming for  `**/i3blocks/config`, probably because there's no builtin hook for that.

As a user this is quite a confusing setup. If kakoune wants to have editorconfig independent configuration for trimming whitespace, I would expect a global option that either enables trimming of all trailing whitespace, or none.

Similar to how there are global tabstop and indentwidth options. And then let the exceptions be configured via either editorconfig, or options local to windows or buffers. Maybe that's not possible, but to me as a user that was what I initially expected, and it took some digging to find out what was actually happening.
